test/integration/secret_test.go

```diff
@@ -0,0 +1,185 @@
+	}
+
+	var m *master.Master
+	s := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		m.Handler.ServeHTTP(w, req)
+	}))
+	defer s.Close()
+
+	m = master.New(&master.Config{
+		EtcdHelper:        helper,
+		KubeletClient:     client.FakeKubeletClient{},
+		EnableLogsSupport: false,
+		EnableUISupport:   false,
+		EnableIndex:       true,
+		APIPrefix:         "/api",
+		Authorizer:        apiserver.NewAlwaysAllowAuthorizer(),
+		AdmissionControl:  admit.NewAlwaysAdmit(),
+	})
+
+	testCases := []string{
+		"v1beta1",
+		"v1beta2",
+	}
+
+	for _, apiVersion := range testCases {
+		deleteAllEtcdKeys()
+		client := client.NewOrDie(&client.Config{Host: s.URL, Version: apiVersion})
+		DoTestSecrets(t, client, apiVersion)
+	}
+}
+
+// DoTestSecrets test secrets for one api version.
+func DoTestSecrets(t *testing.T, client *client.Client, apiVersion string) {
+	// Make a secret object in each of two namespaces.
+	ns1 := "ns1"
+	ns2 := "ns2"
+	secrets := []api.Secret{
+		{
+			ObjectMeta: api.ObjectMeta{
+				Name:      "secret",
+				Namespace: ns1,
+			},
+			Data: map[string][]byte{
+				"data": []byte("value1\n"),
+			},
+		},
+		{
+			ObjectMeta: api.ObjectMeta{
+				Name:      "secret",
+				Namespace: ns2,
+			},
+			Data: map[string][]byte{
+				"data": []byte("value2\n"),
+			},
+		},
+	}
+
+	for _, s := range secrets {
+		defer deleteSecretOrErrorf(t, client, s.Namespace, s.Name)
+		if _, err := client.Secrets(s.Namespace).Create(&s); err != nil {
+			t.Errorf("unable to create test secret %s: %v", s.Name, err)
+		}
+	}
+
+	// Template for pods that use a secret.
+	pod := &api.Pod{
+		ObjectMeta: api.ObjectMeta{
+			Name: "XXX",
+		},
+		Spec: api.PodSpec{
+			Volumes: []api.Volume{
+				{
+					Name: "secvol",
+					VolumeSource: api.VolumeSource{
+						Secret: &api.SecretVolumeSource{
+							Target: api.ObjectReference{
+								Kind:      "Secret",
+								Namespace: "XXX",
+								Name:      "secret",
+							},
+						},
+					},
+				},
+			},
+			Containers: []api.Container{
+				{
+					Name:  "fake-name",
+					Image: "fakeimage",
+					VolumeMounts: []api.VolumeMount{
+						{
+							Name:      "secvol",
+							MountPath: "/fake/path",
+							ReadOnly:  true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// Create a pod to consume secret in its same namespace.
+	pod.ObjectMeta.Name = "uses-same-ns-secret"
+	pod.Spec.Volumes[0].VolumeSource.Secret.Target.Namespace = ns1
+	defer deletePodOrErrorf(t, client, ns1, pod.Name)
+	if _, err := client.Pods(ns1).Create(pod); err != nil {
+		t.Errorf("Failed to create pod: %v", err)
+	}
+
+	// Create a pod to consume secret in a different namespace.
+	pod.ObjectMeta.Name = "uses-other-ns-secret"
+	pod.Spec.Volumes[0].VolumeSource.Secret.Target.Namespace = ns2
+	defer deletePodOrErrorf(t, client, ns1, pod.Name)
+	if _, err := client.Pods(ns1).Create(pod); err != nil {
+		t.Errorf("Failed to create pod: %v", err)
+	}
```

Comment:
Suggest factoring out this cut'n pasted code.  And to illustrate why this is a good idea, I think you have bug due to cut'n'paste in defer deletePod... line.